File: src/leetcode/p0000_0099/p0067_add_binary/AddBinary.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package leetcode.p0000_0099.p0067_add_binary;

public interface AddBinary {
    /**
     * Given two binary strings a and b, return their sum as a binary
     * string.
     *
     * Constraints:
     * Each string does not contain leading zeros except for the zero
     * itself.
     *
     * @param a 1 <= a.length <= 10^4
     *          consist only of '0' or '1' characters.
     * @param b 1 <= b.length <= 10^4
     *          consist only of '0' or '1' characters.
     * @return sum of a and b
     */
    String addBinary(String a, String b);
}

File: src/leetcode/p0000_0099/p0067_add_binary/add_binary.py
from abc import ABC, abstractmethod

class AddBinary(ABC):
    """
    Given two binary strings a and b, return their sum as a binary string.
    """
    @abstractmethod
    def add_binary(self, a: str, b: str) -> str:
        """
        Args:
            a:
                1 <= a.length <= 10^4
                consist only of '0' or '1' characters
            b:
                1 <= b.length <= 10^4
                consist only of '0' or '1' characters
        """
        pass

class AddBinaryLoop(AddBinary):
    def add_binary(self, a: str, b: str) -> str:
        i, j = len(a) - 1, len(b) - 1
        carry = 0
        result = []

        while i >= 0 or j >= 0 or carry:
            total = carry
            if i >= 0:
                total += int(a[i])
                i -= 1
            if j >= 0:
                total += int(b[j])
                j -= 1
            result.append(str(total % 2))
            carry = total // 2
        return ''.join(reversed(result))
